BeefyFeeConfigurator

sircryptomaxi opened the pull request:
Test BeefyFeeConfigurator: [0xa90122Ab61D5416A5d5fd03fCbb8AcDBEcF2c184](https://polygonscan.com/address/0xa90122Ab61D5416A5d5fd03fCbb8AcDBEcF2c184)
Test Vault: [0x49DE18D754004F3ba723fCfeb2C333C77E8351A7](https://polygonscan.com/address/0x49de18d754004f3ba723fcfeb2c333c77e8351a7)

BeefyFeeConfigurator is meant to be referenced by strats before charging fees. A struct containing the total, call, strategist and beefy fees is returned to the strat. If the vault is unknown to the contract then it will return default fees, which is category 0. The keeper or owner can update the category a vault belongs in. Only the owner can change fees in a category, and the total fee cannot be raised more than 9.5%.

The StrategyCommonChefLP is also upgraded to 0.8.0, along with the accompanying contracts. The StratManager and FeeManager have been combined to a new contract StratFeeManager since they are in almost every strat and it is not possible to upgrade without upgrading everything. The StratManager params have been consolidated into a struct to save memory space in the constructor. GasThrottler has also been cloned and upgraded. Most `voter` style strats have been moved to archive since they are needed for now but were causing upgrade problems.

Upgrading to 0.8.0 involves updating OpenZepplin imports, removing `pragma experimental ABIEncoderV2` (if present), removing references to SafeMath, changing `now` to `block.timestamp`.

roman-monk commented:
1) i think BeefyFeeConfigurator should map strategies to feeId (`stratFeeId`) instead of vaults. several reasons: vault works like a tokenizer and pps handler it doesn't really know about fee, strategy works with fees directly it's logical to keep it there, if we upgrade vault there won't be any hidden fee switches as all fees already applied to the strat, if we finally have a multistrat vault we won't need any refactoring for it.
also won't be needed to hardcode this tight logic with external call `IStrategy(msg.sender).vault();` in getFees().
can have a `getVaultFees(address vault)` in FeeConfigurator if it's handy.

2) we need some function for the StratManager to set fee for itself.
in FeeConfig `setStrategyFeeId(uint256 _feeId) external { stratFeeId[msg.sender] = _feeId }`.
in StratManager `function setFeeId(_feeId) onlyManager { feeConfig.setStrategyFeeId(_feeId) }`.
the idea is that strategists can craft vault with specific params by themselves, without asking keepers to change fees, same as they do with callFee or HOD now before transferring ownership.

roman-monk commented on contracts/BIFI/infra/BeefyFeeConfigurator.sol:
```diff
@@ -0,0 +1,143 @@
+
+    // fetch fees for the connected vault when called by a strategy
+    function getFees() external view returns (FeeCategory memory) {
+        address vault = IStrategy(msg.sender).vault();
```
it should be explicit `getFees(address)`. no need to make additional external call, and maybe more important to not hardcode knowledge here that there is a strat, that strat has a vault etc

sircryptomaxi replied:
Implemented this.

roman-monk commented on contracts/BIFI/infra/BeefyFeeConfigurator.sol:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    // set a fee category id for a vault
+    function updateVault(address _vault, uint256 _feeId) external onlyManager {
```
i think would be better to follow set* naming conventions here, as we have in 99% of such cases when we update something.
same as it's described in comment. 
also if it's named `setVaultFeeCategory` then comment just becomes obsolete 🙂

roman-monk replied:
obviously i would recommend to change it to `setStratFeeCategory`

sircryptomaxi replied:
Implemented

MirthFutures commented on contracts/BIFI/infra/BeefyFeeConfigurator.sol:
```diff
@@ -0,0 +1,143 @@
+        uint256 beefy;
+        uint256 call;
+        uint256 strategist;
+        string label;
```
What are we using label for? It might be helpful for others to add a comment in here about what each item represents.

sircryptomaxi replied:
Label is intended as general descriptor of the fee category, inspired by Aave eModes. We don't have to use it, but could be something to use later on. I've added comments to describe each property

MirthFutures commented on contracts/BIFI/infra/BeefyFeeConfigurator.sol:
```diff
@@ -0,0 +1,143 @@
+        string label,
+        bool active
+    );
+    event SetTotalLimit(uint256 totalLimit);
```
Is this an old event? I don't see where it is used.

sircryptomaxi replied:
Old event from when the total fee could be changed, removed that function now so event is obselete

MirthFutures commented on contracts/BIFI/infra/BeefyFeeConfigurator.sol:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    // change keeper
+    function setKeeper(address _keeper) external onlyManager {
```
Should have an event a set function.

sircryptomaxi replied:
Implemented

MirthFutures commented on contracts/BIFI/strategies/Common/StratFeeManager.sol:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    // set new vault (only for strategy upgrades)
+    function setVault(address _vault) external onlyOwner {
```
Should add events for all set functions

sircryptomaxi replied:
Implemented

sircryptomaxi commented:
> 1. i think BeefyFeeConfigurator should map strategies to feeId (`stratFeeId`) instead of vaults. several reasons: vault works like a tokenizer and pps handler it doesn't really know about fee, strategy works with fees directly it's logical to keep it there, if we upgrade vault there won't be any hidden fee switches as all fees already applied to the strat, if we finally have a multistrat vault we won't need any refactoring for it.
>    also won't be needed to hardcode this tight logic with external call `IStrategy(msg.sender).vault();` in getFees().
>    can have a `getVaultFees(address vault)` in FeeConfigurator if it's handy (as additional function for off-chain clients, not as part of strategy<->feeConfig interaction).

Agreed, changed the mapping to strats and removed the tight logic to vaults. Didn't include a `getVaultFees` as this will probably break further on with multi-strat vaults.

> 2. we need some function for the StratManager to set fee for itself.
>    in FeeConfig `setStrategyFeeId(uint256 _feeId) external { stratFeeId[msg.sender] = _feeId }`.
>    in StratManager `function setFeeId(_feeId) onlyManager { feeConfig.setStrategyFeeId(_feeId) }`.
>    the idea is that strategists can craft vault with specific params by themselves, without asking keepers to change fees, same as they do with callFee or HOD now before transferring ownership.

Implemented this.

roman-monk commented:
looks good!
